### PR TITLE
feat: Add select / deselect all to table filter menu #602

### PR DESCRIPTION
**Pull request by aalencar**
Add Select and Deselect all functionality to filter menu

### CODE REVIEW

**Comment by mturoci** [comment]
Thanks @aalencar. Please have a look at `feat/alternative-issue-602` branch for an alternative, non-hacky, solution and let me know what you think

**Comment by aalencar** [comment]
Thanks for providing an alternative, @mturoci !

I like the idea of having a new component for Contextual Menu and not using `setTimeout` to trigger the changes.

But I noticed that the current solution is broken when we have 2 filterable columns. The bug is in `onFilterChange`. Working on a fix.

**Comment by aalencar** [comment]
@mturoci I fixed the issue above, and added other changes to `feat/alternative-issue-602`.

One of the missing fixes was persisting the checkbox status after the contextual menu was closed. For that we needed to set the `checked` property of each item in the `filters` state upon initialisation. This raises a problem: by setting `checked` we now have a controlled component and we need to set ContextualMenu state when the checkboxes are clicked. To achieve that I passed a callback to `onFilterChange` that has the access to ContextualMenu `filters` used to manage the checkbox state.

Let me know what you think of the solution.

**Comment by aalencar** [comment]
I just realized that I might have overlooked the logic for tags. Taking a look at it.

**Comment by mturoci** [comment]
@aalencar if you decided to take the approach proposed in `feat/alternative-issue-602`, can you please also switch branch for this PR? The review will be easier that way.

**Comment by aalencar** [comment]
@mturoci done.

**Comment by mturoci** [comment]
> To achieve that I passed a callback to onFilterChange that has the access to ContextualMenu filters used to manage the checkbox state.

This seems to be way more complex than necessary. IIUC, what you do basically is:

* checkbox is clicked
* call parent `onFilterChange`
* from within top-level component (`onFilterChange`) call the callback to update context menu state

I would suggest a simpler approach, which doesn't require any callback:

* checkbox is clicked
* update menu state right away, then call original `onFilterChange` callback

Or maybe I am missing something?

**Comment by aalencar** [comment]
> update menu state right away, then call original onFilterChange callback

`onFilterChange()` is not the event handler but instead creates a custom handler for each filter.

We can't update right away because we have to update the state from within the handler which is returned by `onFilterChange`, and that's why we also pass a callback to `onFilterChange`.

**Comment by mturoci** [comment]
> We can't update right away because we have to update the state from within the handler which is returned by onFilterChange, and that's why we also pass a callback to onFilterChange.

I still don't understand, why would we have to update the state within the handler and not directly? You can refactor the code, no need to stick with what is there currently.

**Comment by aalencar** [comment]
> why would we have to update the state within the handler and not directly

Because we want to update the ContextualMenu's state on click and that has to be done from within the handler.

**Comment by aalencar** [comment]
@mturoci applied your suggestions. Ready to be reviewed again.

File: ui/src/table.tsx [hunk 1]
```diff
@@ -119,7 +119,7 @@ type WaveColumn = Fluent.IColumn & {
 
 type DataTable = {
   model: Table
-  onFilterChange: (filterKey: S, filterVal: S) => (e?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: B) => void
+  onFilterChange: (filterKey: S, filterVal: S, checked?: boolean) => void
```
Comment (mturoci) [inline ui/src/table.tsx:122]: nit
```suggestion
  onFilterChange: (filterKey: S, filterVal: S, checked?: B) => void
```

File: ui/src/table.tsx [hunk 2]
```diff
@@ -208,7 +217,51 @@ const
...
     return line.substr(1, line.length - 2)
   }).join('\n'),
-  DataTable = ({ model: m, onFilterChange, items, filteredItems, selection, selectedFilters, isMultiple, isSearchable, groups, sort, reset }: DataTable) => {
+  ContextualMenu = ({ onFilterChange, col, listProps, selectedFiltersRef, setFiltersInBulk }: ContextualMenuState) => {
+    const
+      isFilterChecked = (data: S, key: S) => !!selectedFiltersRef.current && selectedFiltersRef.current[data]?.includes(key),
+      [filters, setFilters] = React.useState(col.cellType?.tag
+        ? Array.from(listProps.items.reduce((_filters, { key, text, data }) => {
+          key.split(',').forEach(key => _filters.set(key, { key, text, data }))
+          return _filters
```
Comment (mturoci) [inline ui/src/table.tsx:226]: Didn't check, but it seems like it's missing handling `checked`.

File: ui/src/table.tsx [hunk 3]
```diff
@@ -208,7 +217,51 @@ const
...
+      ),
+      selectAll = () => {
+        setFilters(filters.map(i => ({ ...i, checked: true })))
+        setFiltersInBulk(col.key, filters.map(f => f.key))
+      },
+      deselectAll = () => {
+        setFilters(filters.map(i => ({ ...i, checked: false })))
+        setFiltersInBulk(col.key, [])
+      },
+      getOnFilterChangeHandler = (data: S, key: S) => (_ev?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: boolean) => {
```
Comment (mturoci) [inline ui/src/table.tsx:238]: Odd naming, handlers usually start with on rather than get.

nit
```suggestion
      getOnFilterChangeHandler = (data: S, key: S) => (_ev?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: B) => {
```

File: ui/src/table.tsx [hunk 4]
```diff
@@ -543,26 +579,25 @@ export const
         setGroupByKey(option.key as S)
         initGroups()
       },
-      onFilterChange = React.useCallback((filterKey: S, filterVal: S) => (_e?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: B) => {
+      onFilterChange = React.useCallback((filterKey: S, filterVal: S, checked) => {
```
Comment (mturoci) [inline ui/src/table.tsx:582]: Missing type for checked

File: ui/src/table.tsx [hunk 5]
```diff
@@ -208,7 +217,51 @@ const
     const line = JSON.stringify(row)
     return line.substr(1, line.length - 2)
   }).join('\n'),
-  DataTable = ({ model: m, onFilterChange, items, filteredItems, selection, selectedFilters, isMultiple, isSearchable, groups, sort, reset }: DataTable) => {
+  ContextualMenu = ({ onFilterChange, col, listProps, selectedFiltersRef, setFiltersInBulk }: ContextualMenuState) => {
+    const
+      isFilterChecked = (data: S, key: S) => !!selectedFiltersRef.current && selectedFiltersRef.current[data]?.includes(key),
+      [filters, setFilters] = React.useState(col.cellType?.tag
```
Comment (mturoci) [inline ui/src/table.tsx:223]: Would go for `menuFilters` or something like that so that reader is sure which ones are meant.

**Review by mturoci — CHANGES_REQUESTED** [review]
Thanks @aalencar. A few minor comments. I didn't try your solution yet, will do tmr.

File: ui/src/table.tsx (hunk 3 above)
Comment (aalencar) [reply ui/src/table.tsx:238]: Not a handler, it's a high-order function that creates a handler, hence the name

File: ui/src/table.tsx (hunk 2 above)
Comment (aalencar) [reply ui/src/table.tsx:226]: Good catch.

File: ui/src/table.tsx (hunk 5 above)
Comment (aalencar) [reply ui/src/table.tsx:223]: Indeed, surely more meaningful.

File: ui/src/table.tsx (hunk 3 above)
Comment (mturoci) [reply ui/src/table.tsx]: Agree, fair point.

File: ui/src/table.tsx [hunk 6]
```diff
@@ -543,26 +579,25 @@ export const
         setGroupByKey(option.key as S)
         initGroups()
       },
-      onFilterChange = React.useCallback((filterKey: S, filterVal: S) => (_e?: React.FormEvent<HTMLInputElement | HTMLElement>, checked?: B) => {
+      onFilterChange = React.useCallback((filterKey: S, filterVal: S, checked?: B) => {
         setSelectedFilters(selectedFilters => {
-          const filters = selectedFilters || {}
-
+          let filters: Dict<S[]> | null = null
+          filters = selectedFilters || {}
```
Comment (mturoci) [inline ui/src/table.tsx:585]: What's the reason for this change?

**Review by mturoci — APPROVED** [review]
LGTM @aalencar! Just one minor comment from me. Also, let's merge this after `0.20` is released.

File: ui/src/table.tsx (hunk 6 above)
Comment (aalencar) [reply ui/src/table.tsx]: It was probably a mistake during the refactoring. Reverted.
